circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateRegistry.kt

```diff
@@ -105,29 +108,37 @@ internal class RetainedStateRegistryImpl(retained: MutableMap<String, List<Any?>
     }
   }
 
-  override fun performSave() {
-    val map = retained.toMutableMap()
-    valueProviders.forEach { (key, list) ->
-      if (list.size == 1) {
-        val value = list[0].invoke()
-        if (value != null) {
-          map[key] = arrayListOf<Any?>(value)
-        }
-      } else {
+  override fun saveAll() {
+    val map =
+      valueProviders.mapValues { (_, list) ->
         // If we have multiple providers we should store null values as well to preserve
         // the order in which providers were registered. Say there were two providers.
         // the first provider returned null(nothing to save) and the second one returned
         // "1". When we will be restoring the first provider would restore null (it is the
         // same as to have nothing to restore) and the second one restore "1".
-        map[key] =
-          List(list.size) { index ->
-            val value = list[index].invoke()
-            value
+        list
+          .map { it.invoke() }
+          .onEach { value ->
+            // Allow nested registries to also save
+            if (value is RetainedStateRegistry) {
+              value.saveAll()
+            }
           }
       }
-    }
+
     valueProviders.clear()
-    retained.putAll(map)
+
+    if (map.isNotEmpty()) {
+      retained.putAll(map)
+    }
```

Comment:
Can you put a small comment explaining this snippet for posterity?

Reply:
Done